index.js

```diff
@@ -1,284 +1,161 @@
-      console.warn('not encoding unknown value', input)
-    // FIXME: more checks, including floats!
-    return input
   }
+
+  throw new Error('cannot encode input ' + input)
 }
 
 const decoder = {
-  feed(buf) {
-    const tf = buf.slice(0, 2)
-    const d = buf.slice(2)
-
-    let feedExtension
-    if (tf.equals(CLASSIC_FEED_TF)) feedExtension = '.ed25519'
-    else if (tf.equals(GABBYGR_FEED_TF)) feedExtension = '.ggfeed-v1'
-    else if (tf.equals(BENDYBT_FEED_TF)) feedExtension = '.bbfeed-v1'
-    else throw new Error('Unknown feed: ' + buf)
-
-    const b64part = d.toString('base64')
-    return '@' + b64part + feedExtension
-  },
-
-  message(buf) {
-    const tf = buf.slice(0, 2)
-    const d = buf.slice(2)
-
-    let msgExtension
-    if (tf.equals(CLASSIC_MSG_TF)) msgExtension = '.sha256'
-    else if (tf.equals(GABBYGR_MSG_TF)) msgExtension = '.ggmsg-v1'
-    else if (tf.equals(BENDYBT_MSG_TF)) msgExtension = '.bbmsg-v1'
-    else throw new Error('Unknown msg: ' + buf)
-
-    const b64part = d.toString('base64')
-    return '%' + b64part + msgExtension
+  sigilSuffix(input, type, format) {
+    return [
+      type.sigil || '',
+      input.slice(2).toString('base64'),
+      format.suffix || '',
+    ].join('')
   },
+  bool(input) {
+    if (input.size > 3)
+      throw new Error('boolean BFE must be 3 bytes, was ' + input.size)
+    if (input.slice(2, 3).equals(BOOL.FALSE)) return false
+    if (input.slice(2, 3).equals(BOOL.TRUE)) return true
 
-  blob(buf) {
-    const tf = buf.slice(0, 2)
-    const d = buf.slice(2)
-
-    let blobExtension
-    if (tf.equals(CLASSIC_BLOB_TF)) blobExtension = '.sha256'
-    else throw new Error('Unknown blob ID: ' + buf)
-
-    const b64part = d.toString('base64')
-    return '&' + b64part + blobExtension
-  },
-
-  signature(buf) {
-    const d = buf.slice(2)
-    const b64part = d.toString('base64')
-    return b64part + '.sig.ed25519'
+    throw new Error('invalid boolean BFE')
   },
-
-  box(buf) {
-    const tf = buf.slice(0, 2)
-    const d = buf.slice(2)
-    if (tf.equals(BOX1_TF)) return d.toString('base64') + '.box'
-    else if (tf.equals(BOX2_TF)) return d.toString('base64') + '.box2'
-    else throw new Error('Unknown box: ' + buf)
-  },
-
-  string(buf) {
-    const d = buf.slice(2)
-    return d.toString('utf8')
-  },
-
-  boolean(buf) {
-    const d = buf.slice(2)
-    return d.equals(BOOL_TRUE)
+  string(input) {
+    return input.slice(2).toString('utf8')
   },
 }
 
 function decode(input) {
-  if (Array.isArray(input)) {
-    return input.map(decode)
-  } else if (Buffer.isBuffer(input)) {
+  /* cases we don't decode */
+  if (input === null) return null
+  if (Number.isInteger(input)) return input
+
+  if (Buffer.isBuffer(input)) {
     if (input.length < 2)
       throw new Error(
         'Buffer is missing first two type&format fields: ' + input
       )
-    const t = input.slice(0, 1)
-    const tf = input.slice(0, 2)
```

Comment:
Could you bring back this pattern? It was a nice once, reviewed by others in the team and merged, and your refactor of it seems unrelated to the core purpose of this PR. Repeating `input.slice(0, 2)` everywhere is error prone, not as clear as `tf` and not as performant.

Reply:
yep